File: src/io/github/hulang1024/chinesechess/rule/chess/ChessM.java
package io.github.hulang1024.chinesechess.rule.chess;

import io.github.hulang1024.chinesechess.rule.AbstractChess;
import io.github.hulang1024.chinesechess.rule.ChessPosition;
import io.github.hulang1024.chinesechess.rule.ChineseChessGame;

/**
 * 象
 * @author Hu Lang
 */
public class ChessM extends AbstractChess {
    @Override
    public boolean canGoTo(ChessPosition destPos, ChineseChessGame game) {
        int rowOffset = destPos.row - pos.row;
        int colOffset = destPos.col - pos.col;

        // 不能过河（限制在本方阵地）
        return MoveRules.isInBoundary(game, host, destPos)
            // 只能走“田”
            && Math.abs(rowOffset) == 2 && Math.abs(colOffset) == 2
            // 同时“田”中心不能有棋子
            && (game.chessboard.isEmpty(
                pos.row + (rowOffset > 0 ? +1 : -1), pos.col + (colOffset > 0 ? +1 : -1)));
    }
}

File: client/src/rulesets/chinesechess/rule/ChessM.ts
import ChessPos from "./ChessPos";
import Game from "./Game";
import AbstractChess from "./AbstractChess";
import { isInBoundary, sign } from "./move_rules";
import ChessboardState from "./ChessboardState";

/**
 * 象
 */
export default class ChessM extends AbstractChess {
  /** 是否可过河 */
  public canCrossBoundary = false;

  canGoTo(destPos: ChessPos, chessboardState: ChessboardState, game: Game) {
    const rowOffset = destPos.row - this.pos.row;
    const colOffset = destPos.col - this.pos.col;

    // 只能走“田”
    return Math.abs(rowOffset) == 2 && Math.abs(colOffset) == 2
      // 同时“田”中心不能有棋子
      && (chessboardState.isEmpty(
        this.pos.row + sign(rowOffset),
        this.pos.col + sign(colOffset),
      ))
      // 不能过河（限制在本方阵地）
      && (this.canCrossBoundary || isInBoundary(game, this.host, destPos));
  }

  clone(): ChessM {
    const clone = new ChessM(this.pos, this.host);
    clone.front = this.front;
    clone.canCrossBoundary = this.canCrossBoundary;
    return clone;
  }
}
